### ISSUE TITLE
Inconsistency beetween react-apollo and @apollo/reac-hooks packages

### ISSUE THREAD

**Issue by Vaneste** [apollographql/react-apollo#3373]
Using either ApolloProvider from react-apollo package or @apollo/react-hooks package should provide the same result.

But if you mix both package :  ApolloProvider from react-apollo and say, useQuery from @apollo/react-hooks, you'll end up with a "Could not find "client" in the context or passed in as an option. Wrap the root component in an <ApolloProvider>, or pass an ApolloClient instance in via options." error.

Apollo documentation states that you can use react-apollo OR only @apollo/react-hooks if you just want to use hooks. Implicitely, that means that @apollo/react-hooks is kinf of subset of react-apollo package. So both package should be compatible for using ApolloProvider but this is not the case.

**How to reproduce the issue:**  

Just try to import ApolloProvider from react-apollo and useQuery from @apollo/react-hooks and you'll get the error.

**Version**
Reproduce on : https://codesandbox.io/s/nameless-meadow-jhk6p
Just change the comment in the package.json to get the error.
Package version : 
    "react-apollo": "^2.5.6",
    "apollo-client": "2.6.4",
    "@apollo/react-hooks": "^3.0.0",
Thanks to Alex Banks for his tutorial and the codesandbox sample !

**Comment by jackyef** [comment]
I am having this issue as well. I would expect that importing `ApolloProvider` from either `@apollo/react-common` or `@apollo/react-hooks` would have the same outcome.

**Comment by Ayc0** [comment]
This is normal, the provider changed between the v2 and the v3
react-apollo v3 is using the same as @apollo/react-hooks, but as yours is still in v2, it tries to basically import 2 different libraries.

In general, you should only have one graphql version and one apollo verison.

Bump your `react-apollo` to `"^3.0.0` and it should be fine

**Comment by jackyef** [comment]
@Ayc0 I updated my comment. I am having similar problem but with `@apollo/react-common`

**Comment by Ayc0** [comment]
@jackyef, what are the versions of your packages?

There shouldn't be any difference between importing `ApolloClient` from `@apollo/react-hooks` and `@apollo/react-common` because under the hood, they are the same:  [export statement in the codebase](https://github.com/apollographql/react-apollo/blob/ec4208c20750a8eb605bda2f9f0ffb44fb4723b0/packages/hooks/src/index.ts#L2)

But if you have a mismatched version of `react-apollo`, there **should** be an error because it is not using the same package

**Comment by Ayc0** [comment]
(also look at your lock file to see what is the real versions of every packages and try to clear and reinstall your node_modules)

**Comment by jackyef** [comment]
I am either importing `@apollo/react-common` or `@apollo/react-hooks`. Both are at `3.0.0`. I understand that `@apollo/react-hooks` is just re-exporting `ApolloProvider` from `@apollo/react-common`, that's why I am also confused why this is happening. I will look into it further to see if I can find any more info.

**Comment by jackyef** [comment]
Apparently the problem was because `@apollo/react-hooks` has dependency to `@apollo/react-common^3.0.1`, while my app has dependency to `@apollo/react-hooks@3.0.0` and `@apollo/react-common@3.0.0`. When installing dependencies, I ended up with 2 different version of `@apollo/react-common`, one inside my app `node_modules` and the other one inside `@apollo/react-hooks/node_modules`.

The solution was to follow the dependencies version so there is no duplicates.

tldr:
I changed my package.json deps from:
```
	"@apollo/react-common": "3.0.0",
    "@apollo/react-components": "3.0.0",
    "@apollo/react-hoc": "3.0.0",
    "@apollo/react-hooks": "3.0.0",
    "@apollo/react-ssr": "3.0.0",
    "@apollo/react-testing": "3.0.0",
```
to:
```
    "@apollo/react-common": "^3.0.0",
    "@apollo/react-components": "^3.0.0",
    "@apollo/react-hoc": "^3.0.0",
    "@apollo/react-hooks": "^3.0.0",
    "@apollo/react-ssr": "^3.0.0",
    "@apollo/react-testing": "^3.0.0",
```

**Comment by Vaneste** [comment]
Ok, thanks @Ayc0 & @jackyef ! 
It reminds me good old DLL hell ;)
I think this issue can be close but at least it may be useful for other beginners ;)

**Comment by Ayc0** [comment]
@jackyef, yep, that's why I suggested you to check your lockfile

If node cannot use the same version of a package that already exists in node_modules, it will create a subfolder node_modules with this specific version of the package in it

**Comment by Ayc0** [comment]
That's for this purpose that the `resolutions` field exists in the `package.json`

**Comment by jackyef** [comment]
@Ayc0 Just to clarify, so is the rule of the thumbs on installing react-apollo packages is to only install one of them?

Let's say, I only use hooks, so I only install `@apollo/react-hooks`.
But, if I use HOCs, components and hooks (all three), I should only install `@apollo/react-hoc`, without the others?

**Comment by Ayc0** [comment]
@jackyef you can do whatever you want.

As mentioned [here](https://www.apollographql.com/docs/react/hooks-migration/#bundle-size-note), every package imports the underlying ones.
**But**, they don't import their content (for instance [react-hoc](https://github.com/apollographql/react-apollo/blob/master/packages/hoc/src/index.ts) doesn't export the hooks)

So you can do:
- just use `react-apollo@3` that will export everything
- set one package in the `dependencies` and the other ones in `peerDependencies` (so that you won't re-install them)
- set every one of those in the `dependencies` and set (I think at least) `@apollo/react-common` in the `peerDependencies`
- set every one of those in the `dependencies` and manually check your lock file to ensure that the version that are imported are the right one (most complicated and least resilient solution)

**Comment by jackyef** [comment]
Alright, I see your point, thanks!

**Comment by mplayer78** [comment]
I am still having this issue, even though the only react-apollo I have in my package.json is 		"@apollo/react-hooks": "^3.0.0", my dependencies look like this ```
"dependencies": {
		"@apollo/react-hooks": "^3.0.0",
		"graphql": "^14.5.4",
		"react": "^16.9.0",
		"react-dom": "^16.9.0",
		"react-scripts": "3.1.1"
	},```
and in my yarn.lock I only have "@apollo/react-common@^3.1.0":

Is anyone else still having this issue?

**Comment by cristiandley** [comment]
So the only way to fix this is this way ?

```javascript
import { ApolloProvider } from 'react-apollo';
import { ApolloProvider as ApolloHooks } from '@apollo/react-hooks';

<ApolloProvider client={client}>
        <ApolloHooks client={client}>
          <GlobalStyle />
          <Layout />
        </ApolloHooks>
</ApolloProvider>
```

**Comment by cristiandley** [comment]
```
yarn install v1.16.0
[1/5] 🔍  Validating package.json...
[2/5] 🔍  Resolving packages...
Couldn't find any versions for "@apollo/react-common" that matches "3.1.1"
? Please choose a version of "@apollo/react-common" from this list: (Use arrow keys)
❯ 3.1.0 
  3.1.0-beta.0 
```

**Comment by Ayc0** [comment]
> Seems [3.1.1](https://www.npmjs.com/package/@apollo/react-common) is not a version ?
Yeah, it's the 3.1.0

also my fix was if you were using the v3 everywhere

in your case, you are using `react-apollo` v2, which was defining it's own Provider.
upgrade `react-apollo` to the v3 and it should work

**Comment by akougblenou** [comment]
Hi I think I am having a similar issue.
I am using `apollo-client` and I wanted to use the hooks in my project so I decided to add `@apollo/react-hooks`, but the client always comes back as undefined.
Here is an extract of my dependencies

```js
"resolutions": {
    "@apollo/react-common": "3.1.0"
  },
  "dependencies": {
    "@apollo/react-hooks": "^3.1.1",
    "antd": "3.21.4",
    "apollo-cache-inmemory": "^1.6.3",
    "apollo-client": "^2.6.4",
    "apollo-link-http": "^1.5.16",
...
```

later on my client init looks like this :
```js
import { createHttpLink } from 'apollo-link-http';
import { ApolloClient } from 'apollo-client';

import { InMemoryCache } from 'apollo-cache-inmemory';
import fetch from 'unfetch';
import { GRAPH_QL_URI } from '../constants';

// const httpLink = new HttpLink({
//     uri : GRAPH_QL_URI,
//     fetch,
// });

const httpLink = createHttpLink({ uri : GRAPH_QL_URI, fetch });

export const client = () => new ApolloClient({
    // cache : new InMemoryCache(),
    link : httpLink,
    // uri : GRAPH_QL_URI,
});

```

Finally I call the `ApolloProvider`in my index in such way:
```js
const App = () => (
        <Suspense fallback={<></>}>
            <ApolloProvider client={client}>
               // some stuff happening here
            </ApolloProvider>
        </Suspense>
);

ReactDOM.render(<App />, document.querySelector('#root'));
```
I have added the resolution to `react-common` but it does not seems to solve the issue at all, I might be missing something else ?

**Comment by jackietieu** [comment]
Thanks @Ayc0 , I had the same issue and fixed it thanks to your comments. All of the packages were at 3.1.3, but @apollo/react-testing had a peer dependency of @apollo/react-common. Even though it was the same version as the global react-common, it was throwing an error. Uninstalling react-testing, installing react-common, then re-installing react-testing to remove the peer dependency fixed the issue.

**Comment by Whispers12** [comment]
@jackietieu @akougblenou @Ayc0 @mplayer78 I faced same issue in monorepo

just installing @apollo/react-hooks in package (not root), and it magically works without importing from it (I import only from react-apollo)

I think it's a problem with module resolution or lerna links

**Comment by jamestrenda** [comment]
yeah none of these fixes worked for me. I'm using the same version of react-common, i'm using v3 for everything and I'm still getting this error.

**Comment by Ayc0** [comment]
Way to fix it:
- Check in the `package.json` that each package has the same version as the others
- in you're using yarn, check in `yarn.lock`, if you're using npm, check in `package-lock.json` and here check that each package has the same version as the others (like ^3.0.0 can result to the version 3.0.4)
- use the `resolutions` field of your `package.json` to pin packages
